rfcs/0158-artifact-metadata.md

```diff
@@ -0,0 +1,62 @@
+# RFC 158 - artifact metadata
+* Comments: [#158](https://github.com/taskcluster/taskcluster-rfcs/pull/158)
+* Proposed by: @escapewindow
+
+# Summary
+
+The goal is to provide Artifact Integrity guarantees from the point the worker uploads an artifact, to the point that someone downloads that artifact to use. We can do this by:
+
+1. adding SHA metadata to artifacts in the Queue,
+2. ensuring that that metadata can't be modified once it's written, and
+3. providing a download tool that queries the Queue for an artifact's location and SHA, downloads the artifact, and verifies the downloaded SHA matches the SHA provided by the Queue.
+
+## Motivation
+
+This would improve robustness and security. By storing a SHA to verify on download, we can avoid corrupt downloads. By verifying the read-only SHA before we use the artifact, we can detect malicious tampering of artifacts at rest before we use it for a critical operation.
+
+(This would obsolete the [artifacts section of the Chain of Trust artifact](https://scriptworker.readthedocs.io/en/latest/chain_of_trust.html#chain-of-trust-artifacts), and improve artifact integrity platform-wide.)
+
+# Details
+
+## Adding artifact metadata to the Queue
+
+First, we add a `metadata` dictionary to the `S3ArtifactRequest` type. This is a dictionary to allow for flexibility of usage. The initial known keys would include
+
+```
+ContentFilesize int64 `json:"contentFilesize"`
+ContentSha256 string `json:"contentSha256"`
+ContentSha512 string `json:"contentSha512"`
+```
+
+The sha256 field is required for Artifact Integrity. Mozilla Releng has use cases for all 3 fields, so I'm proposing all 3.
```

Comment:
Why both `ContentSha256` and `ContentSha512`?

Reply:
The above line :)

We could do with just `ContentSha256`. However, when Releng creates our SHA256SUMS and SHA512SUMS files, we have to re-download the files and calculate the shas. Since we have the capability of adding more information here, I thought we might as well support both, and avoid the extra download and sha generation. This is more for efficiency than added security.

Reply:
It occurs to me that generating 2 shas for all artifacts is likely less efficient than re-downloading and re-generating shas for a small subset of artifacts. We could just use `ContentSha256` :)

Reply:
Ok totally. Yeah, it just seems like a recipe for confusion having both.

Reply:
Removed. Now that we removed `ContentSha512`, we likely won't be needing `ContentSha256Signature`, and `ContentFilesize` is optional, it does look more and more like we're making `S3Artifact`s closer to `BlobArtifacts`. We could flatten this dictionary if we don't foresee needing any more metadata, or keep it if we want the option.